### PR TITLE
Proper webhook installation detection

### PR DESCRIPTION
**Pull request by filiptronicek**
## Description

Improves the way we detect if you have webhooks installed on a specific repo. The approach with this PR changes to the following:
1. We check the host if it knows about a webhook for the current Gitpod host. If we lack permissions, we continue to step 2, otherwise, we return the result of this check.
2. (GitHub-specific) we look at the `pushedAt` date to determine when we should've received the last push event and see if we have received some webhook event after that date. If so, we return it.
3. We query the latest webhook events we got for the repo. We go through them one-by-one and if one of the events matches the latest commit SHA of the repo's default branch, we return that event.

## Related Issues

Contributes to ENT-601 as a follow-up for https://github.com/gitpod-io/gitpod/pull/20228

## How to test

https://ft-super-d2d2fa388b8.preview.gitpod-dev.com/repositories

1. Import a repository
2. Enable prebuilds on it (should install webhook)
3. The prebuild settings page should say that the trigger strategy is "Webhook-based"
4. Remove the webhook on the Git host
5. Create a workspace for that repository, wait for it to create
6. Check its prebuild settings again, it should now say "Activity-based" in the trigger strategy section.

/hold

### CODE REVIEW

File: components/server/src/projects/projects-service.ts [hunk 1]
```diff
@@ -554,21 +555,38 @@ export class ProjectsService {
     ): Promise<WebhookEvent | undefined> {
         const context = (await this.contextParser.handle(ctx, user, project.cloneUrl)) as CommitContext;
 
-        const events = await this.webhookEventDb.findByCloneUrl(project.cloneUrl, 1);
+        const events = await this.webhookEventDb.findByCloneUrl(project.cloneUrl, 50);
```
Comment (geropl) [inline components/server/src/projects/projects-service.ts:561]: `prompt("Why especially 50? Could we have chosen a different number as well? What's the tradeoff?") > comment` :orange_heart:

File: components/server/src/prebuilds/github-service.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    constructor(
+        @inject(GitHubRestApi) protected readonly githubApi: GitHubRestApi,
+        @inject(Config) private readonly config: Config,
+        @inject(TokenService) private readonly tokenService: TokenService,
+        @inject(GithubContextParser) private readonly githubContextParser: GithubContextParser,
+    ) {
+        super();
+    }
+
+    async installAutomatedPrebuilds(user: User, cloneUrl: string): Promise<void> {
```
[the remark this answers is not in the document]
Comment (geropl) [reply components/server/src/prebuilds/github-service.ts:51]: Ah, and removed in the 2nd :upside_down_face:

**Review by geropl — APPROVED** [review]
Code LGTM, tested and works! :heavy_check_mark: 

(I had to do a hard reload on the repo config page, but I think that's fine)

File: components/server/src/projects/projects-service.ts (hunk 1 above)
Comment (filiptronicek) [reply components/server/src/projects/projects-service.ts:561]: Thanks for the comment reminder, will add and push 🙏

Comment (filiptronicek) [reply components/server/src/projects/projects-service.ts:564]: Pushed as [`c071537` (#20240)](https://github.com/gitpod-io/gitpod/pull/20240/commits/c071537a9a9c3936358eccdee57aff26ce8f4cd4)

**Pull request merged**